### PR TITLE
Implement :jseval (Issue #334)

### PR DESCRIPTION
**Pull request by Carpetsmoker**
TODO:
- Tests
- Doesn't show errors

### CODE REVIEW

**Comment by The-Compiler** [comment]
I think it'd make more sense to add this to `qutebrowser/browser/commands.py` where there are already utility functions to get the current widget.

**Comment by The-Compiler** [comment]
Regarding showing of errors: I don't think there's a way we can actually get the error via `QWebFrame::evaluateJavaScript` I'm afraid... it returns `None` on errors, but I guess there also are valid JS expressions which don't return anything.

I'd say let's move it to `browser/commands.py` and merge. Tests are rather complex for this currently, so I'm okay with not having tests for this for now.

**Comment by Carpetsmoker** [comment]
You can use the `javaScriptConsoleMessage` override to get the error. I looked into this yesterday, but didn't make any code yet. We need to store the errors somewhere in `javaScriptConsoleMessage`, and then retrieve that in the `jseval` command.

This would also allow the creation of a `:jserrors` command to show all the errors, which would also be a useful command IMHO.

I'd rather do this correct the first time around :-)

**Comment by The-Compiler** [comment]
Hmm. That could work (though when trying it out I didn't actually see anything in the debug logs).

However, wouldn't there be no way to find out whether a given error is from the webpage or from `evaluateJavaScript`? If there isn't, I'd prefer no error handling at all to something which sometimes works, but sometimes not :wink:

**Comment by The-Compiler** [comment]
Oh wow, interesting hack :laughing:

Shouldn't `window.__qute_jseval__` also be deleted when `:jseval` succeeded (i.e. no console message would show up)?

Let me know when it's ready for review.

**Comment by Carpetsmoker** [comment]
Interesting, perhaps, but also not working 100% reliably :-( And I can't think of anything else. The problem is that everything needs to be in a single `evaluateJavaScript` call, otherwise you give other JS code a chance to run (which might generate an error). And puttting anything *after* the JS code the user entered is also not something we want, as `evaluateJavaScript` will reutrn the result of the last statement...

I've reverted this for now. Unless you can think of something I suppose you can merge this.

**Comment by The-Compiler** [comment]
> The problem is that everything needs to be in a single `evaluateJavaScript` call, otherwise you give other JS code a chance to run (which might generate an error).

Are you sure about that? I can't imagine how this would happen, unless QtWebKit runs JS in a separate thread of course...

Another idea I had would be running something like:

```js
window.__qute_jseval__ = true;
returnval = (...);
window.__qute_jseval__ = false;
returnval;
```

or so, with `...` being the thing the user actually wanted to evaluate.

I'm not sure if this works, and if it even can work - but I thought I'd leave the idea here.

**Comment by Carpetsmoker** [comment]
1. jseval -> `__qute_jseval__` = true;
2. jseval code runs, and errors out.
3. We end up in `javaScriptConsoleMessage`, with `__qute_jseval__ == true`
4. We check if it is true, and unset it.

However, between step 3 and 4 other JS code runs, which you can test with:

	<script>
	setInterval(function() {
		if (window.__qute_jseval__) {
			throw new Error('jseval hack failed. Sorry :-( ' + window.__qute_jseval__);
		}
	}, 1);
	</script>

The only only thing I could think of is wrapping the code in a try..else block and running a qutebrowser js function which somehow passes the error message back. But this won't work for all errors...

File: qutebrowser/browser/commands.py [hunk 1]
```diff
@@ -1510,3 +1510,29 @@ def debug_webaction(self, action, count=1):
...
+            # is not trivial...
+            message.info(self._win_id, 'No output or error')
+        else:
+            # The output can be a string, number, dict, array, etc. But *don't*
+            # output too much data, as this will make qutebrowser hang
+            out = str(out)
+            if len(out) > 5000:
+                message.info(self._win_id, out[:5000] + ' [...trimmed...]')
+            else:
+                message.info(self._win_id, out)
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: I suggest using `utils.compact_text()` or `utils.elide()` here - `compact_text` will also compact newlines (I think that's good thing).

**Comment by The-Compiler** [comment]
I see... yeah, it all sounds rather brittle - I'd say let's do it without the error handling for now.

File: qutebrowser/browser/commands.py [hunk 2]
```diff
@@ -1510,3 +1510,29 @@ def debug_webaction(self, action, count=1):
         view = self._current_widget()
         for _ in range(count):
             view.triggerPageAction(member)
+
+    @cmdutils.register(instance='command-dispatcher', scope='window',
+                       maxsplit=0, no_cmd_split=True)
+    def jseval(self, js_code):
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: What about adding a `quiet` flag so the output doesn't get shown?

File: qutebrowser/browser/commands.py (hunk 1 above)
Comment (Carpetsmoker) [inline qutebrowser/browser/commands.py]: Not if you're looking at the terminal output :-)

**Comment by The-Compiler** [comment]
I added `--quiet` and did some changes to the way commands are parsed/handled so the `js_code` argument worked properly.

Thanks!

**Pull request merged**